cornac/models/gru4rec/gru4rec.py

```diff
@@ -0,0 +1,437 @@
+                torch.tensor(np.vstack(m), device=self.G[i].weight_ih.device)
+            )
+            m2 = []
+            m2.append(self._init_numpy_weights((self.layers[i], self.layers[i])))
+            m2.append(self._init_numpy_weights((self.layers[i], self.layers[i])))
+            m2.append(self._init_numpy_weights((self.layers[i], self.layers[i])))
+            self.G[i].weight_hh.set_(
+                torch.tensor(np.vstack(m2), device=self.G[i].weight_hh.device)
+            )
+            self.G[i].bias_hh.set_(
+                torch.zeros((self.layers[i] * 3,), device=self.G[i].bias_hh.device)
+            )
+            self.G[i].bias_ih.set_(
+                torch.zeros((self.layers[i] * 3,), device=self.G[i].bias_ih.device)
+            )
+        self.Wy.weight.set_(
+            torch.tensor(
+                self._init_numpy_weights((self.n_items, self.layers[-1])),
+                device=self.Wy.weight.device,
+            )
+        )
+        self.By.weight.set_(
+            torch.zeros((self.n_items, 1), device=self.By.weight.device)
+        )
+
+    def embed_constrained(self, X, Y=None):
+        if Y is not None:
+            XY = torch.cat([X, Y])
+            EXY = self.Wy(XY)
+            split = X.shape[0]
+            E = EXY[:split]
+            O = EXY[split:]
+            B = self.By(Y)
+        else:
+            E = self.Wy(X)
+            O = self.Wy.weight
+            B = self.By.weight
+        return E, O, B
+
+    def embed_separate(self, X, Y=None):
+        E = self.E(X)
+        if Y is not None:
+            O = self.Wy(Y)
+            B = self.By(Y)
+        else:
+            O = self.Wy.weight
+            B = self.By.weight
+        return E, O, B
+
+    def embed_gru(self, X, H, Y=None):
+        E = self.GE(X, H)
+        if Y is not None:
+            O = self.Wy(Y)
+            B = self.By(Y)
+        else:
+            O = self.Wy.weight
+            B = self.By.weight
+        return E, O, B
+
+    def embed(self, X, H, Y=None):
+        if self.constrained_embedding:
+            E, O, B = self.embed_constrained(X, Y)
+        elif self.embedding > 0:
+            E, O, B = self.embed_separate(X, Y)
+        else:
+            E, O, B = self.embed_gru(X, H[0], Y)
+        return E, O, B
+
+    def hidden_step(self, X, H, training=False):
+        for i in range(self.start, len(self.layers)):
+            X = self.G[i](X, Variable(H[i]))
+            if training:
+                X = self.D[i](X)
+            H[i] = X
+        return X
+
+    def score_items(self, X, O, B):
+        O = torch.mm(X, O.T) + B.T
+        return O
+
+    def forward(self, X, H, Y, training=False):
+        E, O, B = self.embed(X, H, Y)
+        if training:
+            E = self.DE(E)
+        if not (self.constrained_embedding or self.embedding):
+            H[0] = E
+        Xh = self.hidden_step(E, H, training=training)
+        R = self.score_items(Xh, O, B)
+        return R
+
+
+def io_iter(
```

Comment:
Is it something that can be reused for other models?

Reply:
Currently, there is no other model reuse this iteration. We may need to migrate it to `dataset` when there is another model reuse this iteration.